lib/ABI/ModelAlignment.cpp

```diff
@@ -0,0 +1,171 @@
+/// \file ModelAlignment.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ABI/ModelAlignment.h"
+#include "revng/ABI/RuntimeTraitAccessors.h"
+#include "revng/Model/Binary.h"
+
+namespace abi {
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::Type &Type,
+                     model::ABI::Values ABI);
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::QualifiedType &Type,
+                     model::ABI::Values ABI);
+
+template<typename FieldType>
+static RecursiveCoroutine<std::optional<uint64_t>>
+aggregateAlignmentImpl(model::VerifyHelper &VH,
+                       const SortedVector<FieldType> Fields,
+                       model::ABI::Values ABI) {
+  uint64_t Alignment = 0;
+
+  for (const auto &Field : Fields) {
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Field.Type, ABI))
+      Alignment = std::max(Alignment, *A);
+    else
+      rc_return std::nullopt;
+  }
+
+  rc_return Alignment;
+}
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::Type &Type,
+                     model::ABI::Values ABI) {
+  auto MaybeAlignment = VH.size(&Type);
+  if (MaybeAlignment)
+    rc_return MaybeAlignment;
+
+  // This code assumes that the type `Type` is well formed.
+  uint64_t Alignment;
+
+  switch (Type.Kind) {
+  case model::TypeKind::RawFunctionType:
+  case model::TypeKind::CABIFunctionType:
+    // Function prototypes have no size - hence no alignment.
+    Alignment = 0;
+    break;
+
+  case model::TypeKind::PrimitiveType:
+    // The alignment of primitives is simple to figure out based on the abi
+    if (const auto *P = llvm::cast<model::PrimitiveType>(&Type);
+        P->PrimitiveKind == model::PrimitiveTypeKind::Void) {
+      // Make sure void types have no size.
+      if (P->Size != 0)
+        rc_return std::nullopt;
```

Comment:
This is an assertion.
Do not try to be flexible, the model offers invariants: either assert those for extra safety or simply exploit them silently.

Reply:
Done: https://github.com/revng/revng/pull/232/commits/544081721e8bfa61e70661e316bfa247431ca965.